File: aterm-java/src/aterm/ATermFwd.java
package aterm;
import jjtraveler.VisitFailure;


public class ATermFwd implements aterm.Visitor {

    jjtraveler.Visitor any;

    public ATermFwd(jjtraveler.Visitor any) {
        this.any = any;
    }

    public jjtraveler.Visitable visit(jjtraveler.Visitable v) 
    throws jjtraveler.VisitFailure {
	if (v instanceof aterm.Visitable) {
	    return ((aterm.Visitable) v).accept(this);
	}
	    throw new jjtraveler.VisitFailure();
    }

    public aterm.Visitable visitATerm(ATerm arg) throws VisitFailure {
        return (aterm.Visitable) any.visit(arg);
    }

    public aterm.Visitable visitInt(ATermInt arg) throws VisitFailure {
        return visitATerm(arg);
    }

    public aterm.Visitable visitReal(ATermReal arg) throws VisitFailure {
        return visitATerm(arg);
    }

    public aterm.Visitable visitAppl(ATermAppl arg) throws VisitFailure {
        return visitATerm(arg);
    }

    public aterm.Visitable visitList(ATermList arg) throws VisitFailure {
        return visitATerm(arg);
    }

    public aterm.Visitable visitPlaceholder(ATermPlaceholder arg) throws VisitFailure {
        return visitATerm(arg);
    }

    public aterm.Visitable visitBlob(ATermBlob arg) throws VisitFailure {
        return visitATerm(arg);
    }

    public aterm.Visitable visitAFun(AFun fun) throws VisitFailure {
        return fun;
    }
}

File: aterm-csharp/ATerm/Visitors/ATermFwd.cs
using System;
using JJTraveler;

namespace aterm
{
	/// <summary>
	/// Summary description for Visitor.
	/// </summary>
	/// 
	public class ATermFwd : ATermVisitor
	{
		private IVisitor visitor;
		
		public ATermFwd(IVisitor visitor) 
		{
			this.visitor = visitor;
		}

		public virtual IVisitable visit(IVisitable visitable)
		{
			if(visitable is ATermVisitable)
			{
				((ATermVisitable)visitable).accept(this);
				return visitable;
			}
			else
			{
				throw new VisitFailure();
			}
		}
		public virtual void visitATerm(ATerm arg) // throws VisitFailure 
		{
			visitor.visit(arg);
		}

		public virtual void visitInt(ATermInt arg) // throws VisitFailure 
		{
			visitATerm(arg);
		}

		public virtual void visitReal(ATermReal arg) // throws VisitFailure 
		{
			visitATerm(arg);
		}

		public virtual void visitAppl(ATermAppl arg) // throws VisitFailure 
		{
			visitATerm(arg);
		}

		public virtual void visitList(ATermList arg) // throws VisitFailure 
		{
			visitATerm(arg);
		}

		public virtual void visitPlaceholder(ATermPlaceholder arg) // throws VisitFailure 
		{
			visitATerm(arg);
		}

		public virtual void visitBlob(ATermBlob arg) // throws VisitFailure 
		{
			visitATerm(arg);
		}

		public virtual void visitAFun(AFun fun) // throws VisitFailure 
		{
			visitor.visit(fun);
		}
	}
}
